### ISSUE TITLE
3.0 - some [older] discount information may be lost during migration

### ISSUE THREAD

**Issue by robincornett** [easydigitaldownloads/easy-digital-downloads#8643]
### Expected behavior
When I migrate orders from 2.x to 3.0, I expect the data to be consistent.

### Actual behavior
Some orders with discounts applied end up not properly showing/applying the discount after migration:

![image](https://user-images.githubusercontent.com/4853066/119175198-86a72280-ba37-11eb-80d2-0ef79c604f83.png)

The discount applied here is 15% off and the order total should be $41.65, which is what it is in 2.10.

### Steps to reproduce the behavior
I think to reproduce the behavior, you have to have a dataset with orders (with discounts) dating prior to sometime in early December, 2013. In the database I'm looking at, this appears to be approximately when data stops being inaccurate.

### Information (if a specific version is affected):
PHP Version: 7.3.5
EDD Version (or branch): release/3.0
WordPress Version: 5.7.2

Any other relevant information:

Here's the `cart_details` from the payment meta of an affected order:

```
[cart_details] => a:1:{i:0;a:5:{s:4:"name";s:29:"PayPal Pro and PayPal Express";s:2:"id";s:3:"162";s:11:"item_number";a:2:{s:2:"id";s:3:"162";s:7:"options";a:0:{}}s:5:"price";s:5:"49.00";s:8:"quantity";i:1;}}
```

In 3.0, this order appears this way in the database:

In `edd_orders` (limited to relevant columns):
|id |status |subtotal |discount |tax |total |
--- | --- | --- | --- | --- | ---
|107132631 |complete |0.000000000 |0.000000000 |0.000000000 |49.000000000 |

In `edd_order_items`:
|id |order_id |product_id |product_name |price_id |cart_index |quantity |amount |subtotal |discount |tax |total |
--- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | ---
|835 |107132631 |162 |PayPal Pro and PayPal Express — Single Site |0 |0 |1 |49.000000000 |49.000000000 |0.000000000 |0.000000000 |49.000000000 |

And in `edd_order_adjustments`:
|id |object_id |object_type |type_id |type |description |subtotal |tax |total |
--- | --- | --- | --- | --- | --- | --- | --- | ---
|60 |107132631 |order |21916 |discount |TURKEY2012 |0.000000000 |0.000000000 |0.000000000 |

The order item row is nearly correct, although it's not showing the discount. But the amount and subtotal are properly populated, I think because this code accounts for the array key `price` as the `item_price` fallback:

https://github.com/easydigitaldownloads/easy-digital-downloads/blob/677d4f2187938bea4a1680422a84dfd7ff908420/includes/admin/upgrades/v3/class-data-migrator.php#L816-L819

However, the order subtotal/discount are calculated from `$cart_item['subtotal']`, which is present in newer orders, but not in the ones I'm looking at:

https://github.com/easydigitaldownloads/easy-digital-downloads/blob/677d4f2187938bea4a1680422a84dfd7ff908420/includes/admin/upgrades/v3/class-data-migrator.php#L468-L471

This causes the discount to be incorrectly calculated for the order because the `$subtotal` is 0:

https://github.com/easydigitaldownloads/easy-digital-downloads/blob/677d4f2187938bea4a1680422a84dfd7ff908420/includes/admin/upgrades/v3/class-data-migrator.php#L1086-L1087

As far as I can tell, I think we need to tweak the following:
- copy the order item logic (with `price` fallback) back to the order creation calculations, which I think will fix both the subtotal and the discounts on the order, and on the order adjustment.
- account for `$cart_item['discount']` possibly not being present in the cart details array, although if it's not, I'm not sure that it's possible to accurately account for the discount being applied to the individual item, because it could be an excluded item and the discount just applied to something else in the cart.

I _believe_ this is an issue with older orders, so refunds are not a likely scenario, but with this example order, it would be possible to attempt to use the new refund API to issue a refund for more than what the original order actually was, since the customer paid $41.65 but the refund would validate $49 (the gateway would likely balk here).

This also will affect customer/store statistics, I think.

**Comment by robincornett** [comment]
Looked into this some this afternoon and something which might work would be:
- when the `$cart_details` is first defined, run the logic through it that is currently done when building/adding the order items, so the same data is used for the items throughout the function.
- before adding the order to the database, loop through the order discounts and get the totals for them, but store the discount data until adding to the database in its current location.
- calculate the order discount amount while looping through the discounts.
- before adding the order to the database, compare the discount amount from the sums of line 470 (that first loop through the cart items) with what was calculated from the discount objects, and use the larger (if we can assume that the larger is correct) to calculate the order discount amount and order total.

I put something together to try this this afternoon and it seems plausible--wanted to note it here on the issue while it's fresh in my mind.

**Comment by robincornett** [comment]
@ashleyfae I'm closing this as we've resolved it in the PR for #8671.
